src/Functions/serial.cpp

```diff
@@ -0,0 +1,160 @@
+#include <Common/ZooKeeper/ZooKeeper.h>
+#include <Columns/ColumnVector.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionFactory.h>
+#include <Interpreters/Context.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int KEEPER_EXCEPTION;
+}
+
+class FunctionSerial : public IFunction
+{
+private:
+    mutable zkutil::ZooKeeperPtr zk;
+    ContextPtr context;
+
+public:
+    static constexpr auto name = "serial";
+
+    explicit FunctionSerial(ContextPtr context_) : context(context_)
+    {
+        if (context->hasZooKeeper()) {
+            zk = context->getZooKeeper();
+        }
+    }
+
+    static FunctionPtr create(ContextPtr context)
+    {
+        return std::make_shared<FunctionSerial>(std::move(context));
+    }
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 1; }
+    bool isStateful() const override { return true; }
+    bool isDeterministic() const override { return false; }
+    bool isDeterministicInScopeOfQuery() const override { return false; }
+    bool isSuitableForConstantFolding() const override { return false; }
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForNothing() const override { return false; }
+    bool canBeExecutedOnDefaultArguments() const override { return false; }
+    bool isInjective(const ColumnsWithTypeAndName & /*sample_columns*/) const override { return true; }
+    bool hasInformationAboutMonotonicity() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() != 1)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH,
+                "Number of arguments for function {} doesn't match: passed {}, should be 1.",
+                getName(), arguments.size());
+        if (!isStringOrFixedString(arguments[0]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Type of argument for function {} doesn't match: passed {}, should be string",
+                getName(), arguments[0]->getName());
+
+        return std::make_shared<DataTypeInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        if (zk == nullptr)
```

Comment:
I guess we should also check that at most N zk counter nodes are created. Just to make sure there can be no accidental DoS via SQL.

Also, we should have a separate privilege ([comment](https://github.com/ClickHouse/ClickHouse/issues/56228#issuecomment-1789933636)) to run serial. Alternatively, a setting similar to the one for introspection functions https://clickhouse.com/docs/en/sql-reference/functions/introspection may do the job.

Reply:
I can't understand what it means and how to do this